text/0517-io-os-reform.md

```diff
@@ -473,12 +635,289 @@ throughout IO, we can go on to explore the modules in detail.
+impl Write for Vec<u8> { ... }
+```
+
+While efficient, none of these implementations support seeking (via an
+implementation of the `Seek` trait). The implementations of `Read` and `Write`
+for these types is not quite as efficient when `Seek` needs to be used, so the
+`Seek`-ability will be opted-in to with a new `Cursor` structure with the
+following API:
+
+```rust
+pub struct Cursor<T> {
+    pos: u64,
+    inner: T,
+}
+impl<T> Cursor<T> {
+    pub fn new(inner: T) -> Cursor<T>;
+    pub fn into_inner(self) -> T;
+    pub fn get_ref(&self) -> &T;
+}
+
+impl Seek for Cursor<Vec<u8>> { ... }
+impl<'a> Seek for Cursor<&'a [u8]> { ... }
+impl<'a> Seek for Cursor<&'a mut [u8]> { ... }
+
+impl Read for Cursor<Vec<u8>> { ... }
+impl<'a> Read for Cursor<&'a [u8]> { ... }
+impl<'a> Read for Cursor<&'a mut [u8]> { ... }
+
+impl BufferedRead for Cursor<Vec<u8>> { ... }
+impl<'a> BufferedRead for Cursor<&'a [u8]> { ... }
+impl<'a> BufferedRead for Cursor<&'a mut [u8]> { ... }
+
+impl<'a> Write for Cursor<&'a mut [u8]> { ... }
+impl Write for Cursor<Vec<u8>> { ... }
+```
+
+A sample implementation can be found in [a gist][cursor-impl]. Using one
+`Cursor` structure allows to emphasize that the only ability added is an
+implementation of `Seek` while still allowing all possible I/O operations for
+various types of buffers.
+
+[cursor-impl]: https://gist.github.com/alexcrichton/8224f57ed029929447bd
+
+It is not currently proposed to unify these implementations via a trait. For
+example a `Cursor<Rc<[u8]>>` is a reasonable instance to have, but it will not
+have an implementation listed in the standard library to start out. It is
+considered a backwards-compatible addition to unify these various `impl` blocks
+with a trait.
+
+The following types will be removed from the standard library and replaced as
+follows:
+
+* `MemReader` -> `Cursor<Vec<u8>>`
+* `MemWriter` -> `Cursor<Vec<u8>>`
+* `BufReader` -> `Cursor<&[u8]>` or `Cursor<&mut [u8]>`
+* `BufWriter` -> `Cursor<&mut [u8]>`
 
 ### The `std::io` facade
 [The std::io facade]: #the-stdio-facade
 
-> To be added in a follow-up PR.
+The `std::io` module will largely be a facade over `core::io`, but it
+will add some functionality that can live only in `std`.
+
+#### `Errors`
+[Errors]: #error
+
+The `IoError` type will be renamed to `std::io::Error`, following our
+[non-prefixing convention](https://github.com/rust-lang/rfcs/pull/356).
+It will remain largely as it is today, but its fields will be made
+private. It may eventually grow a field to track the underlying OS
+error code.
+
+The `IoErrorKind` type will become `std::io::ErrorKind`, and
+`ShortWrite` will be dropped (it is no longer needed with the new
+`Writer` semantics), which should decrease its footprint. The
+`OtherIoError` variant will become `Other` now that `enum`s are
+namespaced.
+
+The `EndOfFile` variant will be removed in favor of returning `Ok(0)`
+from `read` on end of file. This approach clarifies the meaning of the
```

Comment:
This may also raise a question for `write` as well. For example, the `Write` implementation for `&mut [u8]` currently returns `EndOfFile` when you hit the end of the buffer. I think this is more conventionally known as a "broken pipe" in unix, and it may be a more serious affair than `read()` return 0, although I'm not quite sure just yet.